### ISSUE TITLE
Epic: Personalize workspaces from a user-configurable dotfiles repository

### ISSUE THREAD

**Issue by dnsmichi** [gitpod-io/gitpod#5198]
## Proposal

Provision the environment from a dotfiles repository, e.g. https://gitlab.com/dnsmichi/dotfiles 

May need a special entry point in the .gitpod.yml similar to init/tasks: https://www.gitpod.io/docs/config-gitpod-file 

## Context

Question from @eddiejaoude in the stream at https://www.youtube.com/watch?v=tOrr4r_pNQc 

GitHub Codespaces provides this feature: https://docs.github.com/en/codespaces/customizing-your-codespace/personalizing-codespaces-for-your-account#dotfiles 

## Possible Workarounds

Manual commands in .gitpod.yml and automate the setup of a dotfiles repository. 

---
More [details in the RFC](https://gitpod.notion.site/dotfiles-in-Gitpod-workspaces-b46b8723e9fe4efdbede72daa311961f)

**Comment by dnsmichi** [comment]
FYI @gtsiolis @pawlean

**Comment by eddiejaoude** [comment]
Yes please! This would be amazing 💥

**Comment by bigint** [comment]
Love to see it ✨

**Comment by ghuntley** [comment]
This has been heavily discussed internally and yes, we want to do it or help enable someone in the community to do it! 

<img width="444" alt="2021-08-17_15-17-14" src="https://user-images.githubusercontent.com/127353/129667675-18ecdebb-56c1-406f-be12-90c0dc11242b.png">

[Internal discussion link](https://gitpod.slack.com/archives/C01KGM9DVRC/p1627490845128400?thread_ts=1627475035.128100&cid=C01KGM9DVRC).

**Comment by svenefftinge** [comment]
This needs to be a user setting. 
I think codespaces did a great job on this and we should implement the same semantics as described [here](https://docs.github.com/en/codespaces/customizing-your-codespace/personalizing-codespaces-for-your-account#dotfiles).

The settings should go under a new entry 'Dotfiles'.

**Comment by JanKoehnlein** [comment]
I'd love to have that as well, but it doesn't currently fit into our schedule. Could you bring this up again later?

**Comment by dac09** [comment]
Just leaving a suggestion here: I don't think dotfiles should necessarily be provided in the gitpod.yml - because it's highly personal. I'd much prefer to set the dotfiles on _my_ account only, and not force my team to use my dotfiles too :)

**Comment by eddiejaoude** [comment]
Yep, I would not want my team to have the same dotfiles either, but I would like my dotfiles to load automatically into my gitpod environment - this is similar to CodeSpaces.

**Comment by axonasif** [comment]
> Just leaving a suggestion here: I don't think dotfiles should necessarily be provided in the gitpod.yml - because it's highly personal. I'd much prefer to set the dotfiles on _my_ account only, and not force my team to use my dotfiles too :)

Agreed ;)

**Comment by JanKoehnlein** [comment]
We're currently still swamped, but this issue will be definitely reconsidered when you bring add it to groundwork again later.

**Comment by csweichel** [comment]
Filling in some more detail here.

## Configuration and user flow
Users would point to a repository in their settings on the dashboard. This needs some visual design @gtsiolis.
When a workspace starts, prior to starting the IDE we'd clone the repo and put it in the home directory. Hence, when the IDE starts it'll pull files like `.bashrc` and `.bash_profile`.

## API changes
We could add a `dotfiles_repo` field to the `StartWorkspaceSpec` API. `ws-manager` would translate that field to a `SUPERVISOR_` env var which supervisor would load as part of its workspace specific configuration.

## Implementation
in `supervisor` prior to starting the IDE we'd run `git clone --depth 1` on the dotfile repo, cloning it into `$HOME/.dotfiles`. Once the clone is complete, we'd symlink the files from the dotfiles repo into the home directory. Compared to moving the files into place, we retain the repository hence folks can make changes to their dotfiles and commit them back.

**Comment by roboquat** [comment]
@csweichel: The label(s) `team: ide` cannot be applied. These labels are supported: `team: IDE, team: meta, team: workspace`

<details>

In response to [this](https://github.com/gitpod-io/gitpod/issues/5198#issuecomment-960607912):

>/team IDE


Instructions for interacting with me are available [here](https://github.com/gitpod-io/gitbot/blob/main/plugins/customlabels/README.md). If you have questions or suggestions related to my behavior, please file an issue against the [gitbot](https://github.com/gitpod-io/gitbot/issues/new) repository.
</details>

**Comment by ghuntley** [comment]
> Once the clone is complete, we'd symlink the files from the dotfiles repo into the home directory. 

Let's follow the lead at https://docs.github.com/en/codespaces/customizing-your-codespace/personalizing-codespaces-for-your-account#dotfiles and in our documentation advocate for https://www.gnu.org/software/stow/ (and/or the convention documented in the previous link).

Thus Gitpod looks for the convention, executes the convention and it's the convention that does the symlink (not Gitpod)

> Compared to moving the files into place, we retain the repository hence folks can make changes to their dotfiles and commit them back.

10000% the right choice.

**Comment by bigint** [comment]
This will support both bitbucket and GitLab right?

**Comment by andymac4182** [comment]
Will users need to replicate their dotfiles repository into each provider or will you be able to pick which provider to pull from?

**Comment by iQQBot** [comment]
mark

**Comment by behrica** [comment]
What's about "different dot files" per project ?
Maybe I want a different emacs config per project.

So far, I have one, on my computer. (for all projects)

But maybe dev-containers will bring me to "several emacs configurations, **different per project**.

This might happen, as my current, single, Emacs configuration need to support all languages I work with (and email and an and  Emacs = OS).

I would not need that in Emacs in Gitpod.

**Comment by csweichel** [comment]
More [details in the RFC](https://www.notion.so/gitpod/dotfiles-in-Gitpod-workspaces-b46b8723e9fe4efdbede72daa311961f)

**Comment by andymac4182** [comment]
@csweichel Is that meant to be publicly viewable? I can't see that page.

**Comment by JanKoehnlein** [comment]
Thanks a lot for the PR.

@gtsiolis Please have a look at the dashboard changes. Maybe we need a repo chooser instead of the text field.

**Comment by jldec** [comment]
I tried this PR with my slightly naive dotfiles repo which just tries to set an alias using [`.bash_profile`](https://github.com/jldec/dotfiles/blob/main/.bash_profile)
See https://github.com/jldec/dotfiles

I noticed from the .dotfiles.log that we are symlinking the `.git` directory (bug) and also the `README.md` file (bug?), but not my `.bash_profile` (maybe because there is already a .bash_profile in which case the log should show the failure)

```
gitpod ~ $ cat .dotfiles.log
Cloning into '/home/gitpod/.dotfiles'...
# installation script candidate /home/gitpod/.dotfiles/install.sh is not available
# installation script candidate /home/gitpod/.dotfiles/install is not available
# installation script candidate /home/gitpod/.dotfiles/bootstrap.sh is not available
# installation script candidate /home/gitpod/.dotfiles/bootstrap is not available
# installation script candidate /home/gitpod/.dotfiles/script/bootstrap is not available
# installation script candidate /home/gitpod/.dotfiles/setup.sh is not available
# installation script candidate /home/gitpod/.dotfiles/setup is not available
# installation script candidate /home/gitpod/.dotfiles/script/setup is not available
# echo linking /home/gitpod/.dotfiles/.git/HEAD -> /home/gitpod/.git/HEAD
# echo linking /home/gitpod/.dotfiles/.git/config -> /home/gitpod/.git/config
# echo linking /home/gitpod/.dotfiles/.git/description -> /home/gitpod/.git/description
# echo linking /home/gitpod/.dotfiles/.git/hooks/applypatch-msg.sample -> /home/gitpod/.git/hooks/applypatch-msg.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/commit-msg.sample -> /home/gitpod/.git/hooks/commit-msg.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/fsmonitor-watchman.sample -> /home/gitpod/.git/hooks/fsmonitor-watchman.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/post-update.sample -> /home/gitpod/.git/hooks/post-update.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-applypatch.sample -> /home/gitpod/.git/hooks/pre-applypatch.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-commit.sample -> /home/gitpod/.git/hooks/pre-commit.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-merge-commit.sample -> /home/gitpod/.git/hooks/pre-merge-commit.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-push.sample -> /home/gitpod/.git/hooks/pre-push.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-rebase.sample -> /home/gitpod/.git/hooks/pre-rebase.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/pre-receive.sample -> /home/gitpod/.git/hooks/pre-receive.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/prepare-commit-msg.sample -> /home/gitpod/.git/hooks/prepare-commit-msg.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/push-to-checkout.sample -> /home/gitpod/.git/hooks/push-to-checkout.sample
# echo linking /home/gitpod/.dotfiles/.git/hooks/update.sample -> /home/gitpod/.git/hooks/update.sample
# echo linking /home/gitpod/.dotfiles/.git/index -> /home/gitpod/.git/index
# echo linking /home/gitpod/.dotfiles/.git/info/exclude -> /home/gitpod/.git/info/exclude
# echo linking /home/gitpod/.dotfiles/.git/logs/HEAD -> /home/gitpod/.git/logs/HEAD
# echo linking /home/gitpod/.dotfiles/.git/logs/refs/heads/main -> /home/gitpod/.git/logs/refs/heads/main
# echo linking /home/gitpod/.dotfiles/.git/logs/refs/remotes/origin/HEAD -> /home/gitpod/.git/logs/refs/remotes/origin/HEAD
# echo linking /home/gitpod/.dotfiles/.git/objects/pack/pack-9410ec61f4941fa5eb205496cdd147458b33ffbf.idx -> /home/gitpod/.git/objects/pack/pack-9410ec61f4941fa5eb205496cdd147458b33ffbf.idx
# echo linking /home/gitpod/.dotfiles/.git/objects/pack/pack-9410ec61f4941fa5eb205496cdd147458b33ffbf.pack -> /home/gitpod/.git/objects/pack/pack-9410ec61f4941fa5eb205496cdd147458b33ffbf.pack
# echo linking /home/gitpod/.dotfiles/.git/packed-refs -> /home/gitpod/.git/packed-refs
# echo linking /home/gitpod/.dotfiles/.git/refs/heads/main -> /home/gitpod/.git/refs/heads/main
# echo linking /home/gitpod/.dotfiles/.git/refs/remotes/origin/HEAD -> /home/gitpod/.git/refs/remotes/origin/HEAD
# echo linking /home/gitpod/.dotfiles/.git/shallow -> /home/gitpod/.git/shallow
# echo linking /home/gitpod/.dotfiles/README.md -> /home/gitpod/README.md
```

Beginner Q: If I just wanted to inject aliases with a dotfile, what would be the best way to do that, without overwriting the existing .bashrc in the workspace? Do I need an install script even for something as basic as this?

note: GitHub codespaces does a symlink:  
`.bash_profile -> /workspaces/.codespaces/.persistedshare/dotfiles/.bash_profile`
But the .bash_profile does not appear to be run in each codespaces terminal either. I think this is expected because VS Code terminal are not login shells.

**Comment by jldec** [comment]
As mentioned in https://github.com/gitpod-io/gitpod/pull/7337#issuecomment-1002740006, I suggest shipping the first [MVC PR](https://github.com/gitpod-io/gitpod/pull/7337) as a BETA feature with just a text input UI, and later add a repo selection UI via followup issue #7385.

**Comment by schickling** [comment]
Just wanted drop a (vague) idea here (and maybe it's already possible?):

It would be great if I could persist my VSC settings in a config file/folder which is part of my dotfiles. In my workflow this would replace the MS/GitHub-based settings sync with a more git-centric workflow that could also power GitPod's VSC settings at the same time (related to https://github.com/gitpod-io/gitpod/issues/3733).

On another note related to dotfiles: Please keep Nix user workflows in mind here (cc @ghuntley)

**Comment by loujaybee** [comment]
I've added myself as an assignee here @csweichel to reflect IDE teams support / involvement. However, please remove if confusing / you had another idea of yourself as assignee!

**Comment by gtsiolis** [comment]
Cross-posting from https://github.com/gitpod-io/roadmap/issues/21#issuecomment-1010208934 for visibility:

> Hey everyone! FYI, we've recently shipped the first iteration of the .dotfiles support for workspaces, see gitpod-io/gitpod#5198, gitpod-io/gitpod#7337, and relevant RFC on how this works. 🔖

> It should be relatively straightforward to use this feature. In dashboard /preferences add a repository URL that includes dotfiles. Gitpod will clone and install your dotfiles for every new workspace.

> Follow up issues include 🅰️ adding some relevant documentation (gitpod-io/website#1407 Cc @jldec) and 🅱️ improving the user experience of selecting a repository (gitpod-io/gitpod#7385) from just having a text input for the repository URL. 🗺️

Cc @dnsmichi @behrica @andymac4182 @schickling @eddiejaoude @trumbitta @akosyakov @axonasif @bigint @Milo123459 @patmyron @nberlette @pawlean @jakehamilton @chyld @michael-bowen-sc @hoegertn @jcdea @brams-dev @kr-anurag @vnys @leodido

**Comment by leodido** [comment]
> Cross-posting from [gitpod-io/roadmap#21 (comment)](https://github.com/gitpod-io/roadmap/issues/21#issuecomment-1010208934) for visibility:
> 
> > Hey everyone! FYI, we've recently shipped the first iteration of the .dotfiles support for workspaces, see #5198, #7337, and relevant RFC on how this works. 🔖
> 
> > It should be relatively straightforward to use this feature. In dashboard /preferences add a repository URL that includes dotfiles. Gitpod will clone and install your dotfiles for every new workspace.
> 
> > Follow up issues include 🅰️ adding some relevant documentation ([gitpod-io/website#1407](https://github.com/gitpod-io/website/issues/1407) Cc @jldec) and 🅱️ improving the user experience of selecting a repository (#7385) from just having a text input for the repository URL. 🗺️
> 
> Cc @dnsmichi @behrica @andymac4182 @schickling @eddiejaoude @trumbitta @akosyakov @axonasif @bigint @Milo123459 @PatMyron @nberlette @pawlean @jakehamilton @chyld @michael-bowen-sc @hoegertn @JcdeA @brams-dev @kr-anurag @vnys @leodido

Lovely to see this finally come to life! 💟 
Congrats all!

**Comment by zpthree** [comment]
Looks like the dotfiles repo has to be public?

```
Cloning into '/home/gitpod/.dotfiles'...
remote: Support for password authentication was removed on August 13, 2021. Please use a personal access token instead.
remote: Please see https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/ for more information.
fatal: Authentication failed for 'https://github.com/zpthree/gitpod-dotfiles/'
# dotfile init failed: exit status 128
```

**Comment by gtsiolis** [comment]
@zpthree yes, you're right! We've opened a follow up issue to also support private repositories in https://github.com/gitpod-io/gitpod/issues/7565 and there's already a fix on the way from @iQQBot in https://github.com/gitpod-io/gitpod/pull/7569.

**Comment by zpthree** [comment]
@gtsiolis awesome! looks like my comment got deleted somehow so for anyone seeing this i was just saying that it looks like the dotfiles repo needs to be public.
